text/0030-protocol-messages.md

```diff
@@ -0,0 +1,246 @@
+It should not be used as is but rather be paired with a higher layer - like a type-length-value encoding - and as such
+does not provide any safety check on inputs/outputs.
+
+```rust
+/// A trait describing the behavior of a message.
+trait Message {
+    /// The unique identifier of the message within the protocol.
+    const ID: u8;
+
+    /// Returns the size range of the message as it can be compressed.
+    fn size_range() -> Range<usize>;
+
+    /// Deserializes a byte buffer into a message.
+    /// Panics if the provided buffer has an invalid size.
+    /// The size of the buffer should be within the range returned by the `size_range` method.
+    fn from_bytes(bytes: &[u8]) -> Self;
+
+    /// Returns the size of the message.
+    fn size(&self) -> usize;
+
+    /// Serializes a message into a byte buffer.
+    /// Panics if the provided buffer has an invalid size.
+    /// The size of the buffer should be equal to the one returned by the `size` method.
+    fn into_bytes(self, bytes: &mut [u8]);
+}
+```
+
+**Notes**:
+- `size_range` returns an allowed range for the message size because some parts of some messages can be trimmed. It is
+  used to check if a message coming from a transport layer has a valid size. More details on compression below;
+- `from_bytes`/`into_bytes` panic if incorrectly used, only the following safe TLV module should directly use them;
+- `into_bytes` does not allocate a buffer because the following TLV protocol implies concatenating a header inducing
+  another allocation. Since this is a hot path, a slice of an already allocated buffer for both the header and payload
+  is expected; hence, limiting the amount of allocation to the bare minimum;
+
+## Type-length-value protocol
+
+The [type-length-value](https://en.wikipedia.org/wiki/Type-length-value) module is a safe layer on top of the messages.
+It allows serialization/deserialization to/from a byte buffer ready to be sent/received to/from a transport layer by
+prepending or reading a header containing the type and length of the payload.
+
+### Header
+
+```rust
+/// A header for the type-length-value encoding.
+struct Header {
+    /// Type of the message.
+    message_type: u8,
+    /// Length of the message.
+    message_length: u16,
+}
+```
+
+### Methods
+
+```rust
+/// Deserializes a TLV header and a byte buffer into a message.
+/// * The advertised message type should match the required message type.
+/// * The advertised message length should match the buffer length.
+/// * The buffer length should be within the allowed size range of the required message type.
+fn tlv_from_bytes<M: Message>(header: &Header, bytes: &[u8]) -> Result<M, TlvError> {
+    ...
+}
+
+/// Serializes a TLV header and a message into a byte buffer.
+fn tlv_into_bytes<M: Message>(message: M) -> Vec<u8> {
+    ...
+}
+```
+
+## Messages
+
+Since the messages are all different one from another, there is no construction method in the `Message` trait. All the
```

Comment:
Suggestion: "Since the various types of messages are constructed with different kind of data, there can't be singular constructor signature in the `Message` trait."